autorest/azure/devicetoken_test.go

```diff
@@ -0,0 +1,222 @@
+package azure
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+	"testing"
+	"time"
+
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/mocks"
+	"github.com/Azure/go-autorest/autorest/to"
+)
+
+const (
+	TestResource = "SomeResource"
+	TestClientID = "SomeClientID"
+)
+
+const MockDeviceCodeResponse = `
+{
+	"device_code": "10000-40-1234567890",
+	"user_code": "ABCDEF",
+	"verification_url": "http://aka.ms/deviceauth",
+	"expires_in": "900",
+	"interval": "0"
+}
+`
+
+const MockDeviceTokenResponse = `{
+	"access_token": "accessToken",
+	"refresh_token": "refreshToken",
+	"expires_in": "1000",
+	"expires_on": "2000",
+	"not_before": "3000",
+	"resource": "resource",
+	"token_type": "type"
+}
+`
+
+func TestDeviceCodeIncludesResource(t *testing.T) {
+	sender := mocks.NewSender()
+	sender.EmitContent(MockDeviceCodeResponse)
+	sender.EmitStatus("OK", 200)
+	client := &autorest.Client{Sender: sender}
+
+	code, err := InitiateDeviceAuth(client, TestClientID, TestResource)
+	if err != nil {
+		t.Errorf("unexpected error initiating device auth")
+	}
+
+	if *code.Resource != TestResource {
+		t.Errorf("InitiateDeviceAuth failed to stash the resource in the DeviceCode struct")
+	}
+}
+
+func TestDeviceCodeReturnsErrorIfSendingFails(t *testing.T) {
+	sender := mocks.NewSender()
+	sender.EmitErrors(1)
+	sender.SetError(fmt.Errorf("this is an error"))
+	client := &autorest.Client{Sender: sender}
+
+	_, err := InitiateDeviceAuth(client, TestClientID, TestResource)
+	if err == nil {
+		t.Errorf("failed to get err") // (expecting Error occurred while requesting Device Authorization Code)
+	}
+}
+
+func TestDeviceCodeReturnsErrorIfBadRequest(t *testing.T) {
+	sender := mocks.NewSender()
+	sender.EmitStatus("Bad Request", 400)
+	client := &autorest.Client{Sender: sender}
+
+	_, err := InitiateDeviceAuth(client, TestClientID, TestResource)
+	if err == nil {
+		t.Errorf("failed to get err") // (expecting Request for Device Authorization Code resulted in Status)
+	}
```

Comment:
Whoops, didn't realize it was an exported one. Those should certainly be tested. Sure, there are other error tests that do not refer to exported errors. Do we want assertions on substrings for those? Or just these since they're exported?